lib/os/p4wq.c

```diff
@@ -0,0 +1,209 @@
+/*
+ * Copyright (c) 2020 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include <logging/log.h>
+#include <sys/p4wq.h>
+#include <wait_q.h>
+#include <ksched.h>
+#include <init.h>
+
+LOG_MODULE_REGISTER(p4wq);
+
+struct device;
+
+static void set_prio(struct k_thread *th, struct k_p4wq_work *item)
+{
+	__ASSERT_NO_MSG(!IS_ENABLED(CONFIG_SMP) || !z_is_thread_queued(th));
+	th->base.prio = item->priority;
+	th->base.prio_deadline = item->deadline;
+}
+
+static bool rb_lessthan(struct rbnode *a, struct rbnode *b)
+{
+	struct k_p4wq_work *aw = CONTAINER_OF(a, struct k_p4wq_work, rbnode);
+	struct k_p4wq_work *bw = CONTAINER_OF(b, struct k_p4wq_work, rbnode);
+
+	if (aw->priority != bw->priority) {
+		return aw->priority > bw->priority;
+	}
+
+	if (aw->deadline != bw->deadline) {
+		return aw->deadline - bw->deadline > 0;
+	}
+
+	return (uintptr_t)a < (uintptr_t)b;
+}
+
+/* Slightly different semantics: rb_lessthan must be perfectly
+ * symmetric (to produce a single tree structure) and will use the
+ * pointer value to break ties where priorities are equal, here we
+ * tolerate equality as meaning "not lessthan"
+ */
+static inline bool item_lessthan(struct k_p4wq_work *a, struct k_p4wq_work *b)
+{
+	if (a->priority > b->priority) {
+		return true;
+	} else if ((a->priority == b->priority) &&
+		   (a->deadline != b->deadline)) {
+		return a->deadline - b->deadline > 0;
+	}
+	return false;
+}
+
+static FUNC_NORETURN void p4wq_loop(void *p0, void *p1, void *p2)
+{
+	ARG_UNUSED(p1);
+	ARG_UNUSED(p2);
+	struct k_p4wq *queue = p0;
+	k_spinlock_key_t k = k_spin_lock(&queue->lock);
+
+	while (true) {
+		struct rbnode *r = rb_get_max(&queue->queue);
+
+		if (r) {
+			struct k_p4wq_work *w
+				= CONTAINER_OF(r, struct k_p4wq_work, rbnode);
+
+			rb_remove(&queue->queue, r);
+			w->thread = _current;
+			sys_dlist_append(&queue->active, &w->dlnode);
+			set_prio(_current, w);
+
+			k_spin_unlock(&queue->lock, k);
+			w->handler(w);
+			k = k_spin_lock(&queue->lock);
+
+			/* Remove from the active list only if it
+			 * wasn't resubmitted already
+			 */
+			if (w->thread == _current) {
+				sys_dlist_remove(&w->dlnode);
+				w->thread = NULL;
+			}
+		} else {
+			z_pend_curr(&queue->lock, k, &queue->waitq, K_FOREVER);
+			k = k_spin_lock(&queue->lock);
+		}
+	}
+}
+
+void k_p4wq_init(struct k_p4wq *queue)
+{
+	memset(queue, 0, sizeof(*queue));
+	z_waitq_init(&queue->waitq);
+	queue->queue.lessthan_fn = rb_lessthan;
+	sys_dlist_init(&queue->active);
+}
+
+void k_p4wq_add_thread(struct k_p4wq *queue, struct k_thread *thread,
+			k_thread_stack_t *stack,
+			size_t stack_size)
+{
+	k_thread_create(thread, stack, stack_size,
+			p4wq_loop, queue, NULL, NULL,
+			K_HIGHEST_THREAD_PRIO, 0, K_NO_WAIT);
```

Comment:
I guess my last lack of understanding this week: I don't see the new thread being added anywhere to the queue? How will `k_p4wq_submit()` find it?

Reply:
The thread adds itself to the waitq in p4wq_loop(), which is its entry function.

Reply:
indeed, thanks!